### ISSUE TITLE
Signal uses 100% CPU after waking from long sleep

### ISSUE THREAD

**Issue by tw-hx** [signalapp/Signal-Android#11398]
- [ x ] I have searched open and closed issues for duplicates
- [ x ] I am submitting a bug report for existing functionality that does not work as intended
- [ x ] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [ x ] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
After waking the phone from a long sleep e.g. overnight, Signal starts using 100% of one CPU thread in the background, even without opening the app, until killed. You can otherwise use the app as normal.

### Steps to reproduce
- Use phone normally first thing in the day e.g. listening to a podcast, after the phone has been asleep for at least several hours. Screen can be off, you don't have to open Signal.

**Actual result:**
- Realise it's getting awfully warm in your pocket, and the battery is running low.
- Check the org.thoughtcrime.securesms process is using 100% CPU.
- Force stop and restart Signal. It then operates normally for several hours at least.

**Expected result:**
- Signal uses only the expected background CPU.

### Device info
**Device:** Samsung Galaxy S9+
**Android version:** 11.0 (LineageOS 18.1) with no Google apps installed. The problem started around the time I upgraded from Android 10 (LineageOS 17.1). This might be a ROM problem (sorry in advance if false alarm!), but no other app behaves this way.
**Signal version:** 5.13.8, has happened for that last couple of stable builds. I'm running a self-compiled release build and am happy to compile a debug build if that'd help.

### Link to debug log
https://debuglogs.org/1e2a8005e077dcd79b3306972a1a431336986bd9645ad817cb2c9e5468501f2c
Weird thing is that I am a member of a couple of groups, and it's repeatedly checking the 3 member group to see if a group call is active? I'm not calling or expecting a call.

Thanks if you can shed some light on this!

**Comment by LoveIsGrief** [comment]
Might explain #11604

**Comment by tw-hx** [comment]
Update on this: It's one group causing an infinite loop. I'm a member of several groups but the one affected group is the only one in which I've partaken in a group call previously.

Receiving a message/notification (to any group, not just that group) appears to trigger it at random. Debug logs show repeated calls beginning with "peek group call", then it uses 100% CPU until you either restart Signal or _click on the affected group name in the conversation list_, loading the group messages, at which point CPU use goes back to normal.

I can't get perfetto tracing working on my phone sadly, and system tracing in developer settings crashes.

Looking over the Signal source though, `jobs/GroupCallPeekJob.java` seems suspicious as it's only constraint on that job running is the websocket being drained. Perhaps it needs a `.setMaxAttempts()` in its parameter builder?

**Comment by tw-hx** [comment]
OK, have done some more investigation. It's reproducible outside a sleeping context.

This is triggered by particular groups; I'm a member of 3 groups and when I tap from the conversation list into the group for 2 of them, Signal immediately starts using 100% CPU; the third group is totally unaffected and acts like a normal conversation. This CPU use persists as long as the group is visible; if you go into any group settings or "all media" screens it's still 100%+. When you exit back to the conversation list, CPU usage returns to normal UNLESS the group is open in a chat bubble, in which case 100% usage continues until the bubble is closed.

Looking over the logfile more (which is tricky as this straight up breaks the debug log scrolling functionality), it seems the ultimate reason is the the ConversationActivity is continually being paused and resumed in an infinite loop. The repeated lines all seem to come from onResume:

https://github.com/signalapp/Signal-Android/blob/d3b061c6a4f870fd5edb1d15def484d849a7e039/app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationActivity.java#L587

Has anyone else managed to replicate this? It's not obvious using Signal at all, you have to have a CPU monitor open in the background. Seems like it must stem from something in the database for particular groups that triggers an infinite pause/resume loop.

**Comment by porst17** [comment]
I think I am having the same issue (Samsung Galaxy S10, LineageOS 18.1, Gapps). Running `htop` reveals Signal is running at 100% on one CPU core after receiving messages in certain groups. I sent a system bug report to @greyson-signal . I hope it helps at least confirming this issue and possibly figuring out why this is happening.

**Comment by garymoon** [comment]
I can confirm several aspects of this. I'm also on LineageOS, but have no GApps or MicroG. I have never participated in a group call, but two of three group chats I'm in trigger this issue - 100% CPU usage visible in top/htop that is alleviated when I leave the group chat.

I don't love that the only option appears to be posting a debug log publicly (can't save to file or copy/paste), and the offer to review it before posting rings a little hollow when it's thousands of lines long. Is there an alternative?

**Comment by LoveIsGrief** [comment]
If I'm not mistaken, the debug logs are only collected for the running process, not older ones. That means, force closing Signal should reset the logs, which should make them smaller and possible to review.

It should also be possible to send an email directly to the Signal developers (their emails are public IIRC, check the commits).

**Comment by garymoon** [comment]
> If I'm not mistaken, the debug logs are only collected for the running process, not older ones. That means, force closing Signal should reset the logs, which should make them smaller and possible to review.

You are correct, however it would still be public which remains undesirable.

> It should also be possible to send an email directly to the Signal developers (their emails are public IIRC, check the commits).

This would be my preference, but the contents of the report cannot be copied.

**Comment by tw-hx** [comment]
Somewhere between 5.28.10 and v5.29.7 this problem looks to have resolved for me. Potentially it was https://github.com/signalapp/Signal-Android/commit/5a51544caec713b9570a8aa9f922c0feeebcfb3d which fixed an infinite loop in text rendering? Anyone else still have CPU use issues with group chats?

**Comment by porst17** [comment]
Yes, seems like the issue is solved. Can not reproduce this anymore.

**Comment by cody-signal** [comment]
Based on two users reporting issue fixed, will close. Reopen or comment if the issue still persists for other folks.

**Comment by LoveIsGrief** [comment]
I still have an issue with battrt consumption. How do I verify it's not due to this? How do I verify Signal's CPU usage to confirm (or no)t that this issue has been resolved?

**Comment by LoveIsGrief** [comment]
Well, I'm  not sure the problem is solved as Signal still likes to use 10-20% of my battery just by existing in the background.

No method has been explained or described how to confirm that this bug was actually fixed, in order to rule it out as the cause.
Could it be provided?

**Comment by porst17** [comment]
I had a Signal group that triggered the issue every time I checked. After receiving a message in the group, the device started to heat up. I connected to the device via ADB and checked CPU load via `htop`. Signal was running at 100% CPU. This does not happen anymore on my Samsung Galaxy S10 running official Lineage OS 18.1.

I recommend connecting via ADB and checking the CPU load of the Signal app `org.thoughtcrime.securesms`.
